hashes/src/lib.rs

```diff
@@ -220,6 +220,16 @@ pub trait GeneralHash: Hash {
         }
         Self::from_engine(engine)
     }
+
+    /// Hashes the entire contents of the `reader`.
+    #[cfg(all(feature = "std", feature = "bitcoin-io"))]
+    fn hash_reader<R: std::io::Read>(reader: &mut R) -> Result<Self, std::io::Error>
+    where <Self as GeneralHash>::Engine: std::io::Write
+    {
+        let mut engine = Self::engine();
+        std::io::copy(reader, &mut engine)?;
+        Ok(Self::from_engine(engine))
```

Comment:
Why is the logic duplicated?

Reply:
Can you be more specific, which bit of duplication are you concerned about?

Reply:
Same lines in trait impl and inherent impl. The inherent impl can just call into trait one.

Reply:
Ah, well noticed. I added a patch up front that removes the same sort of duplication in `hash` and `hash_byte_chunks`. I couldn't remove it in the `sha256t` module because of the `Tag`, I didn't try too hard though.